artemis-commons/src/main/java/org/apache/activemq/artemis/api/core/SimpleString.java

```diff
@@ -575,4 +468,64 @@ public void getChars(final int srcBegin, final int srcEnd, final char[] dst, fin
          dst[d++] = (char) (low | high);
       }
    }
+
+   public static final class ByteBufSimpleStringPool extends AbstractByteBufPool<SimpleString> {
+
+      private static final int UUID_LENGTH = 36;
+
+      private final int maxLength;
+
+      public ByteBufSimpleStringPool() {
+         this.maxLength = UUID_LENGTH;
+      }
+
+      public ByteBufSimpleStringPool(final int capacity, final int maxCharsLength) {
+         super(capacity);
+         this.maxLength = maxCharsLength;
+      }
+
+      @Override
+      protected boolean isEqual(final SimpleString entry, final ByteBuf byteBuf, final int offset, final int length) {
+         if (entry == null) {
+            return false;
+         }
+         return entry.equals(byteBuf, offset, length);
+      }
+
+      @Override
+      protected boolean canPool(final ByteBuf byteBuf, final int length) {
+         assert length % 2 == 0 : "length must be a multiple of 2";
+         final int expectedStringLength = length >> 1;
+         return expectedStringLength <= maxLength;
+      }
+
+      @Override
+      protected SimpleString create(final ByteBuf byteBuf, final int length) {
+         return readSimpleString(byteBuf, length);
+      }
+   }
+
+   public static final class StringSimpleStringPool extends AbstractPool<String, SimpleString> {
+
+      public StringSimpleStringPool() {
+         super();
+      }
+
+      public StringSimpleStringPool(final int capacity) {
+         super(capacity);
+      }
+
+      @Override
+      protected SimpleString create(String value) {
+         return toSimpleString(value);
+      }
+
+      @Override
+      protected boolean isEqual(SimpleString entry, String value) {
+         if (entry == null) {
+            return false;
+         }
+         return entry.toString().equals(value);
```

Comment:
Sorry to be pedantic about this method, but I think would be better to define into `SimpleString` something like this:
```
   public boolean isContentEquals(final String value) {
      if (this.str != null) {
         return this.str.equals(value);
      } else {
         if (value == null)
            return false;
         for (int i = 0, size = value.length(); i < size; i++) {
            if (value.charAt(i) != this.charAt(i))
               return false;
         }
         return true;
      }
   }
```
And uses it into `isEqual(SimpleString entry, String value)`.
The reason is the same as before: if a new value has any collision with an existing entry (ie `isEqual` return `false`), the pool will replace the old entry with a new one creating an additional `String` and hence nullifying any perf benefit of using `String` and and adding new work to the GC for nothing.

Reply:
PLease do step through constructing a SimpleString with a String, and then calling toString you will notice no string is created as SimpleString holds the original

Reply:
Let me try to explain better what is the case I'm talking about.
Consider these steps:
1) getOrCreate(a)
2) getOrCreate(b) where hashCode(a) === hashCode(b) but a != b

The call to getOrCreate(b) will create an instance (if not already created) of simpleStringA.toString() to compare with b, but due to the collision of a and b both a and its new toString will be thrown away because b will take its place.
I hope that this explain better the case I was worried about: with a small pool and/or often changing strings having such collisions is the common case and we will have a doubled allocation if compared to the unpooled case.

Reply:
@michaelandrepearce I haven't read that:
> str = string;

Sorry @michaelandrepearce I've missed that part, you're totally right!

Reply:
no worries, in that case could i remove the last change i just did which was adding your method to appease you :) , as i added what you asked, and rather we don't add changes and code for fun.

Ill raise another PR, as it looks like you merged.